src/primitives/hrp.rs

```diff
@@ -0,0 +1,180 @@
+// Written by Clark Moody and the rust-bitcoin developers.
+// SPDX-License-Identifier: MIT
+
+//! Provides an `Hrp` type that represents the human-readable part of a bech32 encoded string.
+//!
+//! > The human-readable part, which is intended to convey the type of data, or anything else that
+//! > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::borrow::Cow;
+use core::fmt;
+use core::iter::FusedIterator;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Hrp<'a> {
+    /// Lowercase human-readable part.
+    Lower(&'a str),
+    /// Uppercase human-readable part.
+    Upper(&'a str),
+    /// All digit human-readable part.
+    Digit(&'a str),
```

Comment:
I guess we will need it if we want to fix the case-checking bug in the top-level parsing function.... but I don't think we should bother, I think we should have a `check_characters` function that does validity checks on the whole string rather than checking for consistency with the HRP.

Once we do that I'm not sure that there's any value in having distinct `Lower`/`Upper` variants at all...we should just unconditionally use bitshifts to convert case when it's needed.